src/ripple/app/paths/impl/PaySteps.cpp

```diff
@@ -370,6 +374,308 @@ toStrand (
+            return {temBAD_PATH, Strand{}};
+    }
+
+    Issue curIssue = [&]
+    {
+        auto const& currency =
+            sendMaxIssue ? sendMaxIssue->currency : deliver.currency;
+        if (isXRP (currency))
+            return xrpIssue ();
+        return Issue{currency, src};
+    }();
+
+    auto hasCurrency = [](STPathElement const pe)
+    {
+        return pe.getNodeType () & STPathElement::typeCurrency;
+    };
+
+    std::vector<STPathElement> normPath;
+    // reserve enough for the path, the implied source, destination,
+    // sendmax and deliver.
+    normPath.reserve(4 + path.size());
+    {
+        normPath.emplace_back(
+            STPathElement::typeAll, src, curIssue.currency, curIssue.account);
+
+        if (sendMaxIssue && sendMaxIssue->account != src &&
+            (path.empty() || !path[0].isAccount() ||
+             path[0].getAccountID() != sendMaxIssue->account))
+        {
+            normPath.emplace_back(sendMaxIssue->account, boost::none, boost::none);
+        }
+
+        for (auto const& i : path)
+            normPath.push_back(i);
+
+        auto const lastCurrency =
+            (*boost::find_if(boost::adaptors::reverse(normPath), hasCurrency))
+                .getCurrency();
+        if (lastCurrency != deliver.currency)
+            normPath.emplace_back(
+                boost::none, deliver.currency, deliver.account);
+
+        if (!((normPath.back().isAccount() &&
+               normPath.back().getAccountID() == deliver.account) ||
+              (dst == deliver.account)))
+        {
+            normPath.emplace_back(deliver.account, boost::none, boost::none);
+        }
+
+        if (!normPath.back().isAccount() ||
+            normPath.back().getAccountID() != dst)
+        {
+            normPath.emplace_back(dst, boost::none, boost::none);
+        }
+    }
+
+    auto const strandSrc = normPath.front().getAccountID ();
+    auto const strandDst = normPath.back().getAccountID ();
+
+    Strand result;
+    result.reserve (2 * normPath.size ());
+
+    /* A strand may not include the same account node more than once
+       in the same currency. In a direct step, an account will show up
+       at most twice: once as a src and once as a dst (hence the two element array).
+       The strandSrc and strandDst will only show up once each.
+    */
+    std::array<boost::container::flat_set<Issue>, 2> seenDirectIssues;
+    // A strand may not include the same offer book more than once
+    boost::container::flat_set<Issue> seenBookOuts;
+    seenDirectIssues[0].reserve (normPath.size());
+    seenDirectIssues[1].reserve (normPath.size());
+    seenBookOuts.reserve (normPath.size());
+    auto ctx = [&](bool isLast = false)
+    {
+        return StrandContext{view, result, strandSrc, strandDst, isLast,
+            ownerPaysTransferFee, seenDirectIssues, seenBookOuts, j};
+    };
+
+    for (int i = 0; i < normPath.size () - 1; ++i)
```

Comment:
fixed `size_t` issue.

Note: `normPath` will always be at least two. If `size` were one, we'd still be OK (the loop would not run at all). If `size` were zero, we'd have a bug, as the `size()-1` would wrap around.